src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -84,5 +224,187 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+
+            LockFileTargetLibrary node = default;
+
+            // Find first target node that matches current
+            foreach (LockFileTargetLibrary lib in graph.Libraries)
+            {
+                if (lib.Type == LibraryType.Package.Value
+                    && string.Equals(lib.Name, current.PackageIdentity.Id, StringComparison.OrdinalIgnoreCase))
+                {
+                    node = lib;
+                    break;
+                }
+            }
+
+            /* 
+             * && ((current.HasAllowedVersions && current.AllowedVersions.Satisfies(lib.Version)) ||
+                        (current.PackageIdentity.HasVersion && current.PackageIdentity.Version.Equals(lib.Version)))
+             */
+
+            if (node != default)
+            {
+                visited.Add(current.PackageIdentity); // visited
+
+                // Lookup Transitive Origins Cache
+                TransitiveEntry cachedEntry;
+                if (!TransitiveOriginsCache.TryGetValue(current.PackageIdentity.Id, out cachedEntry))
+                {
+                    cachedEntry = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
+                    {
+                        [fxRidEntry] = new List<PackageReference>()
+                    };
+                }
+
+                if (!cachedEntry.ContainsKey(fxRidEntry))
+                {
+                    cachedEntry[fxRidEntry] = new List<PackageReference>();
+                }
+
+                if (!cachedEntry[fxRidEntry].Contains(top))
+                {
+                    cachedEntry[fxRidEntry].Add(top);
+                }
+
+                // Upsert Transitive Origins Cache
+                TransitiveOriginsCache[current.PackageIdentity.Id] = cachedEntry;
+
+                foreach (PackageDependency dep in node.Dependencies.ToList()) // Casting to list to prevent backing allocations
+                {
+                    // Create PackageReference object as a data-model based on dependency
+                    var pkgChild = new PackageReference(
+                        identity: new PackageIdentity(dep.Id, dep.VersionRange.MinVersion),
+                        targetFramework: fxRidEntry.Framework,
+                        userInstalled: false,
+                        developmentDependency: false,
+                        requireReinstallation: false,
+                        allowedVersions: dep.VersionRange);
+
+                    if (!visited.Contains(pkgChild.PackageIdentity))
+                    {
+                        MarkTransitiveOrigin(top, pkgChild, graph, visited, fxRidEntry, token);
+                    }
+                }
+            }
+        }
+
+        internal static TransitivePackageReference MergeTransitiveOrigin(PackageReference currentPackage, TransitiveEntry transitiveEntry)
+        {
+            var transitiveOrigins = new SortedSet<PackageReference>(PackageReferenceMergeComparer);
```

Comment:
How this sorting works here? Is it what we really wanted?
Let's say we had following dependency tree for transitive dependency `Deprecated`.
`E -> A ->  C -> Deprecated.`

Then after sorting you get list `A, C, E`.  It's little bit different from how actual tree looks.

Reply:
I am assuming Transitive Origin will be sorted by package Id in ascending order. So, I'm ordering by packageId asc.